src/Collapse.js

```diff
@@ -1,234 +1,169 @@
-      return this.renderFixed();
-    }
-
-    const renderStatic = this.renderStatic;
-    const {height} = this.state;
-    const currentStringHeight = parseFloat(height).toFixed(1);
-
-    if (height > -1 && renderStatic) {
-      this.renderStatic = false;
+    const isAutosize = this.state.action === STABLE && this.state.to;
+    const extraStyle = {};
+    if (isAutosize) {
+      extraStyle.height = 'auto';
+    } else {
+      extraStyle.overflow = 'hidden';
+      if (this.state.action === WAIT && !this.state.to) {
+        extraStyle.height = 0;
+        extraStyle.overflow = 'hidden';
+      }
     }
 
-    // Cache Content so it is not re-rendered on each animation step
-    const content = <HeightReporter onHeightReady={this.onHeightReady}>{children}</HeightReporter>;
 
-    if (renderStatic) {
-      const newStyle = isOpened ? {height: 'auto'} : {overflow: 'hidden', height: 0};
-
-      if (!isOpened && height > -1) {
-        if (!keepCollapsedContent) {
-          return null;
-        }
-
-        return (
-          <div style={{height: 0, overflow: 'hidden', ...style}} {...props}>
-            {content}
-          </div>
-        );
-      }
+    const motionProps = this.state.action === STABLE ? {
+      // When waiting, instantly jump to the position
+      defaultStyle: {height: this.state.to},
+      style: {height: this.state.to}
+    } : {
+      // Otherwise, animate
+      defaultStyle: {height: this.state.from},
+      style: {height: spring(this.state.to, {precision: SPRING_PRECISION, ...springConfig})}
+    };
 
-      // <Motion> to prevent loosing input after causing this component to rerender
-      return (
-        <Motion
-          defaultStyle={{height: Math.max(0, height)}}
-          style={{height: Math.max(0, height)}}
-          onRest={onRest}>
-          {() =>
-            <div style={{...newStyle, ...style}} {...props}>{content}</div>
-          }
-        </Motion>
-      );
-    }
 
     return (
-      <Motion
-        defaultStyle={{height: Math.max(0, height)}}
-        onRest={onRest}
-        style={{height: this.getMotionHeight(height)}}>
-        {st => {
-          this.height = stringHeight(st.height);
-
-          // TODO: this should be done using onEnd from ReactMotion, which is not yet implemented
-          // See https://github.com/chenglou/react-motion/issues/235
-          if (!isOpened && this.height === '0.0') {
-            if (!keepCollapsedContent) {
-              return null;
-            }
-            return (
-              <div style={{height: 0, overflow: 'hidden', ...style}} {...props}>{content}</div>
-            );
-          }
-
-          const newStyle = isOpened && this.height === currentStringHeight ? {height: 'auto'} : {
-            height: st.height, overflow: 'hidden'
-          };
-
-          return <div style={{...newStyle, ...style}} {...props}>{content}</div>;
-        }}
+      <Motion {...motionProps} onRest={this.onRest}>
+        {({height}) => (
```

Comment:
(in the real use case from which this feature request came, we might have to copy Collapse to our code base and start maintaining it)